broadcast-box: init at 0-unstable-2025-06-04

JManch opened the pull request:
## Description of changes

Adds broadcast-box package and module.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

JManch committed:
broadcast-box: init at 2025-06-04

JManch committed:
nixos/tests/broadcast-box: init

fpletz commented on nixos/modules/services/video/broadcast-box.nix:
```diff
@@ -0,0 +1,189 @@
+      description = mdDoc ''
+        Path to the SSL key for Broadcast Box's HTTP server.
+      '';
+    };
```
I know that you're simply copying the name of the environment variables here but I feel like putting the SSL options in the namespace `https.` would make this a bit easier to understand. No requirement, though, just a thought since the naming comes from upstream.

JManch replied:
That's a good idea, I'll group them under `https` along with the `redirect` option.

fpletz commented on pkgs/by-name/br/broadcast-box/package.nix:
```diff
@@ -0,0 +1,62 @@
+    homepage = "https://github.com/Glimesh/broadcast-box";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jmanch ];
+    platforms = [ "x86_64-linux" ];
```
Any specific reason why this wouldn't build on other architectures or platforms? I think `platforms.linux` should be a safe bet and `platforms.unix` would probably also work thanks to Go.

fpletz commented on pkgs/by-name/br/broadcast-box/ignore-env-file.patch:
```diff
@@ -0,0 +1,32 @@
+-		if err := godotenv.Load(envFileDev); err != nil {
+-			log.Fatal(err)
+-		}
+-	}
```
I agree with removing the error message since they would be very misleading in the NixOS module. I'm not sure if we should remove support for loading `.env` files completely without any notice though. People can still run the service binarty without the NixOS module and might be opting to use it with an `.env` file as it is documented upstream.

Maybe we should still load the production env file if `$APP_ENV` is set to `production`? And additionally don't try load any `.env` file at all if `$APP_ENV` is set to an empty string or `nixos` and set this in the NixOS module?

JManch replied:
I agree that being able to respect `$APP_ENV` and load `.env` if they exist would be a nicer solution. Rather than removing that entire block of code in the patch I'll just swap the `log.Fatal(err)` lines with `log.Println(err)` in the case where `APP_ENV == "production"`. This way, broadcast box will only exit due to no `.env` file when `APP_ENV` has not been defined. This should help inform users (running the package outside of the NixOS module) that environment variables need to be set whilst still allowing the program to run in a service without a `.env` file.

Only downside is that the service in the NixOS module will log a harmless `".env.production file does not exist"` error on start-up. I could fix this by patching a special case for `$APP_ENV == "nixos"` but I'm not sure it's worth it?

JManch replied:
On second thought that error message was bothering me so I've modified the patch to check for `$APP_ENV == "nixos"`. Let me know if you'd rather avoid the extra patching.

JManch replied:
Broadcast box have now made `.env.production` the default instead of `.env.development` when `$APP_ENV` is undefined. To keep things simple I will just add a special case for `$APP_ENV == "nixos"` that doesn't attempt to load any `.env` files. 

Running broadcast box standalone without `$APP_ENV` set to `nixos` will require a `.env.production` file in the cwd. Otherwise, broadcast box quits with a `".env.production: no such file"` error. This matches upstream documentation so I think it's acceptable behaviour.

fpletz reviewed (commented):
Thanks a lot! I was also thinking about packaging this.

All in all, you're doing a lot of things right here. Splitting the commits was done perfectly, code formatting is on point and in general you thought of all the small issues some reviewers here tend to be picky about.

Since this is your first PR in nixpkgs, I commend and thank you for doing your research on how to contribute! :+1: Welcome to nixpkgs! :tada: 

I have added a few comments and welcome your feedback on on those.

I'm also going to deploy this and might have more feedback soon.

JManch commented on nixos/modules/services/video/broadcast-box.nix:
```diff
@@ -0,0 +1,189 @@
+      description = mdDoc ''
+        TCP Port the HTTP server will listen on.
+      '';
+    };
```
Grouping with `http` is a good idea, I'll do that. If I understand the docs correctly, Broadcast Box actually can send WebRTC traffic over TCP with the `TCP_MUX_ADDRESS` variable (which I will add as an option). What do you think about something like:
```
http.listenAddress
http.port

udpMux.port
tcpMux.listenAddress
tcpMux.port
```
That way it clearly differentiates the http vs WebRTC addresses and ports. This would also allow me to add options for `UDP_MUX_PORT_(WHEP|WHIP)` as `udpMux.whep.port` and `udpMux.whip.port`.

JManch commented on nixos/modules/services/video/broadcast-box.nix:
```diff
@@ -0,0 +1,189 @@
+        if `ssl.redirect` is enabled.
+
+        If `udp.port` is `null`, it is assigned a random port, so this option
+        will **not** open it in the firewall.
```
I'm a little hesitant to say it opens on all interfaces because it's technically possible for a user's firewall config to alter this behavior (as I do in mine). Also I had a look at how other services document the option and they don't seem to mention it.

I'll add a note about the interface filter.

JManch commented on nixos/modules/services/video/broadcast-box.nix:
```diff
@@ -0,0 +1,189 @@
+      default = true;
+      description = mdDoc ''
+        Whether Broadcast Box should be started automatically.
+      '';
```
Auto start is just a personal thing. On some hosts I like to automatically run services on boot whilst on others I like to start/stop manually. I don't feel strongly about it so I'll remove it (it's easy functionality for users to add themselves anyway).

JManch commented:
Thank you for the kind words and the suggestions! I've replied to your comments and made changes where necessary.

I've also added some more options so now an option exists for every documented environment variable. I think I will still keep the `extraEnv` option in case Broadcast Box adds new options in the future.

I tried testing the WebRTC-over-TCP functionality but no matter what I set `TCP_MUX_ADDRESS` to, I get "address already in use". I'm assume this is user-error rather than a problem with the NixOS module or Broadcast Box, but I'd be interested to hear if you experience the same issue.

I've changed the wording of port options so that, rather than mentioning in every port option that random ports will not open the firewall, the `openFirewall` option now just a sentence warning that randomly assigned ports will not be opened. Hope this is ok.

Once again, thanks for the help and I look forward to hearing your feedback.

JManch commented:
I've made some changes to the module including:

- Systemd hardening
- The status API is now disabled by default because it exposed stream keys
- The `nat1To1Ip` and `includePublicIpInNat1To1Ip` options have been renamed to `nat.ip` and `nat.autoConfigure` (which is now `true` by default to match the upstream docker compose config)
- Added the new `tcpMux.force` option
- Added an example to the `stunServers` option
- Various wording improvements

I've also exposed the NixOS test in the package with `passthru.tests`.

JManch commented:
@fpletz are you still interested in reviewing this?

JManch commented:
I've rewritten and greatly simplified the module. Rather than creating an option for every upstream environment variable, I've replaced them with a single `settings` option that accepts any variable in uppercase snake case. The submodule only defines the necessary options for determining which ports needs to be opened.

Scrumplex commented on pkgs/by-name/br/broadcast-box/package.nix:
```diff
@@ -0,0 +1,69 @@
+}:
+let
+  name = "broadcast-box";
+  version = "2024-07-15";
```
As per https://github.com/NixOS/nixpkgs/blob/3540ebe5534a9e7f38b3c782dba8a03eaeae7303/pkgs/README.md#versioning this should start with `0-unstable-`, unless upstream starts tagging commits.

```suggestion
  version = "0-unstable-2024-07-15";
```

Scrumplex commented on pkgs/by-name/br/broadcast-box/package.nix:
```diff
@@ -0,0 +1,69 @@
+    runHook preInstall
+
+    mkdir -p $out/share
+    cp -r ${frontend}/build/* $out/share
```
I would prefer this to be like this:
```suggestion
    cp -r $frontend/build/* $out/share
```

Note that `${}` is evaluated by Nix so the final value of `installPhase` would contain a static reference to the store path of the `frontend` derivation. If you wish to use `${}` as part of shell syntax, you need to escape the $ with two single quotes like this:

```suggestion
    cp -r ''${frontend}/build/* $out/share
```

This allows users to override the frontend package if they ever wish to do so, like this:

```nix
pkgs.broadcast-box.overrideAttrs (prevAttrs: {
  frontend = prevAttrs.frontend.override {...};
});
```

Scrumplex commented on nixos/modules/services/video/broadcast-box.nix:
```diff
@@ -0,0 +1,239 @@
+    ++ optional (settings.UDP_WHIP_PORT != null) settings.UDP_WHIP_PORT;
+
+  tcpPorts =
+    [ httpPort ]
```
I am sure some people would like to run this behind a reverse proxy, so maybe this should be gated behind an additional option?

Scrumplex reviewed (commented):
Other than my small remarks, this seems to be in an excellent shape ^^

I will try to deploy this to my server as an alternative to OvenMediaEngine and see how well it works

Scrumplex commented:
Seems to work great in my testing! I was able to use a reverse proxy for the web UI/API without any issues :D

JManch commented:
Thank you for the review! I've implemented your suggestions and updated the package.

I've replaced the `HTTP_ADDRESS` variable with `address` and `port` options so
that should make it a bit easier to run behind a reverse proxy.

Scrumplex commented:
> I've replaced the `HTTP_ADDRESS` variable with `address` and `port` options so that should make it a bit easier to run behind a reverse proxy.

Oh I probably wasn't quite clear. What I meant specifically was having an option that would determine if the HTTP port should be opened in the firewall. I think having the environment variable as before was totally fine, though I also don't mind `address` and `port` ^^

JManch committed:
nixos/broadcast-box: init

JManch commented:
I've added separate `openFirewall` and `web.openFirewall` options

Scrumplex commented on nixos/modules/services/video/broadcast-box.nix:
```diff
@@ -0,0 +1,271 @@
+    };
+  };
+
+  options.services.broadcast-box = {
```
nit:

`options` usually come first in a module file.

phaer commented:
@JManch Happy to merge this, once the conflict with all-tests is removed (i.e. rebased).

(nit: You both, alias some functions from lib at the beginning of the file as well as using a fully qualified name for others (i.e. `lib.last`). I think it would be nice to have it consistent per file at least. :))

JManch commented:
@phaer thanks for looking at this. I've bumped the package, cleaned up a few things, and aliased all of the lib functions, so should be good to go :)
